Support SDHC driver for Renesas RA8 devices

khoa-nguyen-18 opened the pull request:
Add support for SDHC driver running on Renesas RA8 
- boards: renesas: Add support for sdhc
- drivers: disk: Add support for sdhc driver
- dts: renesas: Add support for SDHC node.
- dts: bindings: Add support for sdhc driver.
- tests: drivers: disk: disk_performance: Add support for RA8.
- tests: drivers: disk: disk_access: Add support for RA8.

danieldegrasse reviewed (changes requested):
Thank you for the contribution, but this driver is implementing the disk API directly for an SD host controller. New SD host controllers need to be implemented using the SDHC API: https://docs.zephyrproject.org/latest/hardware/peripherals/sdhc.html.

Take a look at the drivers within `drivers/sdhc` for some examples of how this would be implemented.

khoa-nguyen-18 commented:
> Thank you for the contribution, but this driver is implementing the disk API directly for an SD host controller. New SD host controllers need to be implemented using the SDHC API: https://docs.zephyrproject.org/latest/hardware/peripherals/sdhc.html.
> 
> Take a look at the drivers within `drivers/sdhc` for some examples of how this would be implemented.

We noticed a similar implementation for the disk driver on the SDMMC IP in `drivers\disk\sdmmc_stm32.c`. If we want to support SDMMC, do we have to implement it in `drivers/sdhc`, instead of `drivers/disk`?

danieldegrasse commented:
> We noticed a similar implementation for the disk driver on the SDMMC IP in `drivers\disk\sdmmc_stm32.c`. If we want to support SDMMC, do we have to implement it in `drivers/sdhc`, instead of `drivers/disk`?

Yes, you do- I'm sorry that the STM32 implementation made things unclear. This implementation is present from before the SDHC layer was introduced, and needs to be transitioned to use the SDHC class instead of the disk driver layer. New implementations should not be based off it

quytranpzz commented:
@danieldegrasse : thanks for your information. We have implemented the driver using the SDHC API instead of the disk driver layer. Please take a look.

danieldegrasse commented on drivers/sdhc/sdhc_renesas_ra.c:
```diff
@@ -0,0 +1,718 @@
+		ra_cmd.data = (uint8_t *)data->data;
+		ra_cmd.sector_start = data->block_addr;
+		ra_cmd.sector_count = data->blocks;
+		timeout_cfg = data->timeout_ms;
```
This isn't used, but needs to be. You need to respect the timeout configuration in some way- typically this done by waiting for an command completion signal (either via polling or interrupt) until the timeout expires. You should also be waiting for the command timeout, *then* the data timeout. The data timeout does not take precedence over the command timeout when present.

quytranpzz replied:
updated to wait the command timeout by `sdhi_command_send_wait`.

danieldegrasse commented on drivers/sdhc/sdhc_renesas_ra.c:
```diff
@@ -0,0 +1,718 @@
+			}
+		}
+
+#if SDHI_CFG_SD_SUPPORT_ENABLE
```
Just a question- where does this definition come from, the HAL? Could we use 
```
if (IS_ENABLED(SDHI_CFG_SD_SUPPORT_ENABLE))
```
here, or would that cause a compile error?

quytranpzz replied:
Yes. I tried to use if (IS_ENABLED(SDHI_CFG_SD_SUPPORT_ENABLE)) but it causes a compile error.

danieldegrasse commented on drivers/sdhc/sdhc_renesas_ra.c:
```diff
@@ -0,0 +1,718 @@
+		/* Verify card is back in transfer state after write */
+		while (!sdmmc_event.transfer_completed) {
+			/* Wait for transfer. */
+			k_sleep(K_USEC(10));
```
Why not use a semaphore here, rather than polling on this event to be updated by the callback?

danieldegrasse commented on drivers/sdhc/sdhc_renesas_ra.c:
```diff
@@ -0,0 +1,718 @@
+			case SDHC_TIMING_DDR50:
+			case SDHC_TIMING_DDR52:
+				/* Enable DDR mode */
+				LOG_INF("DDR mode enabled");
```
Not clear here- is there supposed to be some change to the hardware applied? Or is that done elsewhere?

quytranpzz replied:
Sorry for the mistake here. Currently, we don't support DDR mode in our hardware so I removed it.

danieldegrasse commented on samples/subsys/fs/fs_sample/boards/ek_ra8d1.conf:
```diff
@@ -0,0 +1 @@
+CONFIG_DISK_DRIVER_SDMMC=y
```
This should be set automatically, no need to do it here

quytranpzz replied:
Thanks for letting me know. I removed it (for the other tests/samples also)

danieldegrasse commented on samples/subsys/fs/fs_sample/boards/ek_ra8d1.overlay:
```diff
@@ -0,0 +1,19 @@
+};
+
+&sdhc1 {
+	status = "okay";
```
Any reason we can't just put this in the base board devicetree and avoid overlays for every SDHC sample?

quytranpzz replied:
This is a mistake. We already added this on board devicetree, so this is redundant, I have removed it. Thank you

danieldegrasse commented on boards/renesas/ek_ra8d1/ek_ra8d1.dts:
```diff
@@ -42,6 +42,7 @@
 
 	aliases {
 		led0 = &led1;
+		sdhc0 = &sdhc1;
```
Need to add `-sdhc` to the boards' test YAML so the SDHC tests will actually run

quytranpzz replied:
Added `-sdhc` into the boards' yaml.

danieldegrasse commented:
Thanks for the rework here, this looks much better. General request, please split the commits up a bit more. Something like this:
- Commit updating HAL
- Commit adding driver code and DTS binding
- Commit adding DTS entry to RA8 SOC
- Commit adding board support for the SDHC to the ek_r8d1
- Commit adding test support for the board

quytranpzz committed:
drivers: sdhc: support SDHC driver on Renesas RA8

This is initial commit to support SDHC driver on Renesas RA8
with r_sdhi modules

Signed-off-by: Tran Van Quy <quy.tran.pz@renesas.com>

quytranpzz committed:
dts: renesas: add dts entry node to support sdhc on Renesas RA8

Add node on RA8 dts to support sdhc as channel 0 and channel 1

Signed-off-by: Tran Van Quy <quy.tran.pz@renesas.com>

quytranpzz committed:
boards: renesas: Add board support for SDHC driver

Add board support for SDHC driver on EK-RA8M1, EK-RA8D1 and MCK-RA8T1

Signed-off-by: Tran Van Quy <quy.tran.pz@renesas.com>

quytranpzz commented on drivers/sdhc/sdhc_renesas_ra.c:
```diff
@@ -0,0 +1,718 @@
+
+struct sdhc_ra_config {
+	const struct pinctrl_dev_config *pcfg;
+
```
dropped. thank you

quytranpzz commented on drivers/sdhc/sdhc_renesas_ra.c:
```diff
@@ -0,0 +1,718 @@
+	.transfer_error = false,
+};
+
+void r_sdhi_callback(sdmmc_callback_args_t *p_args)
```
thanks for your comment. I defined it in INIT function as r_sdhi_callback_##index to use for each sd host controllers

quytranpzz commented on drivers/sdhc/sdhc_renesas_ra.c:
```diff
@@ -0,0 +1,718 @@
+
+LOG_MODULE_REGISTER(sdhc_renesas_ra, CONFIG_SDHC_LOG_LEVEL);
+
+extern fsp_err_t r_sdhi_command_send(sdhi_instance_ctrl_t *p_ctrl, uint32_t command,
```
You're correct. These are internal functions defined in HAL, at [here](https://github.com/renesas/hal_renesas/blob/support_renesas_ra8_sdhc/drivers/ra/fsp/src/r_sdhi/r_sdhi.c#L244) (.c file). For some reason, I follow the structure in our HAL (as FSP support) and extern to use these here.

quytranpzz commented:
@danieldegrasse : I split commits as your comment. Please take a look. Thank you

danieldegrasse commented on drivers/sdhc/sdhc_renesas_ra.c:
```diff
@@ -0,0 +1,736 @@
+			goto end;
+		}
+		/* SDResponseR2 are bits from 8-127, first 8 MSBs are reserved */
+		p_csd_reg.reg.sdrsp10 = priv->sdmmc_ctrl.p_reg->SD_RSP10;
```
So why do we need additional handling of the CSD here? Typically the SD host controller does not care which command is being sent, beyond the type (IE if the command contains data or not)

danieldegrasse commented on drivers/sdhc/sdhc_renesas_ra.c:
```diff
@@ -0,0 +1,736 @@
+		break;
+
+	default:
+		ret = -ENOTSUP;
```
Is this intentional? IE are other SD opcodes not supported by this driver?

quytranpzz replied:
You're correct, other SD opcodes not supported by this driver. I added a comment here for more clearly

danieldegrasse commented on drivers/sdhc/sdhc_renesas_ra.c:
```diff
@@ -0,0 +1,736 @@
+	ret = err_fsp2zep(fsp_err);
+
+	if (ret < 0) {
+		LOG_INF("\r\nR_SDHI_Open error: %d\r\n", fsp_err);
```
Don't need the `\r` here, and likely no leading `\n` either

quytranpzz replied:
Thank you. I removed it

danieldegrasse commented on dts/arm/renesas/ra/ra8/ra8x1.dtsi:
```diff
@@ -557,6 +557,36 @@
+			compatible = "renesas,ra-sdhc";
+			channel = <0>;
+			bus-width = <4>;
+			block-size = <512>;
```
I'm not clear what the `block-size` property is here for- is this a property of the hardware itself? IE can the hardware only support cards of this block size?

quytranpzz replied:
I use `block-size = <512>;` as a default value when init the driver in `R_SDHI_Open`,  it belongs to configs of struct `st_sdmmc_cfg`. In source, I get it by `DT_INST_PROP`, and user can modify it by overlay this device tree by themself

danieldegrasse replied:
See my comment here: https://github.com/zephyrproject-rtos/zephyr/pull/79862/files#r1854354437. I do not think we should be setting this property on the hardware- it is a property of each transfer request, not a fixed value.

danieldegrasse commented on samples/subsys/fs/fs_sample/boards/ek_ra8d1.overlay:
```diff
@@ -0,0 +1,12 @@
+ */
+
+&sdhc1 {
+	sdmmc {
```
These same overlays appear all over for each test. Can we just define the SDMMC node within the base board devicetree?

quytranpzz replied:
Yes, I added it into board devicetree and remove overlay for tests also. Thank you

quytranpzz commented on drivers/sdhc/sdhc_renesas_ra.c:
```diff
@@ -0,0 +1,736 @@
+
+LOG_MODULE_REGISTER(sdhc_renesas_ra, CONFIG_SDHC_LOG_LEVEL);
+
+extern fsp_err_t r_sdhi_transfer_write(sdhi_instance_ctrl_t *const p_ctrl, uint32_t block_count,
```
I added a comment before these functions, thank you.

quytranpzz commented on drivers/sdhc/sdhc_renesas_ra.c:
```diff
@@ -0,0 +1,736 @@
+	sdhimmc_dma_req_isr();
+}
+
+static struct sdmmc_ra_event sdmmc_event[SDHI_MAX_CHANNEL_SUPPORT];
```
Yes, I moved it into `struct sdhc_ra_priv`. Thank you

quytranpzz commented on drivers/sdhc/sdhc_renesas_ra.c:
```diff
@@ -0,0 +1,736 @@
+void sdhimmc_card_isr(void);
+void sdhimmc_dma_req_isr(void);
+
+static void ra_sdmmc_accs_isr(const void *parameter)
```
Thanks for your comment.
Yes, we use the `R_FSP_IsrContextGet` to determine the instance of SDHI peripheral. There might be a bit of concern here, but we intend to reuse our funtions in HAL and minimize the modifications

quytranpzz commented on drivers/sdhc/sdhc_renesas_ra.c:
```diff
@@ -0,0 +1,736 @@
+
+	/* Reset semaphore */
+	k_sem_reset(&sdmmc_event[priv->channel].transfer_sem);
+	k_sem_take(&priv->thread_lock, K_MSEC(ra_cmd.timeout_ms));
```
Sorry for the mistake here. I intend to use this semaphore as protection for the data (I updated the semaphore takes as K_FOREVER). The command timeout is used in `sdhi_command_send_wait`  as your concern below. Not all commands have the event so we can't wait for the callback comes with a semaphore.

danieldegrasse commented on drivers/sdhc/sdhc_renesas_ra.c:
```diff
@@ -0,0 +1,746 @@
+	case SD_READ_SINGLE_BLOCK:
+	case SD_READ_MULTIPLE_BLOCK:
+		/* Configure the transfer interface for reading.*/
+		fsp_err = r_sdhi_transfer_read(&priv->sdmmc_ctrl, ra_cmd.sector_count,
```
Ok, I *think* I understand what is going on here (please correct me if I'm wrong):
- the `sector_count` is being used as the size of block for the DTC (which seems like a DMA controller) to transfer
- the `block_size` is being used as the number of blocks for the DTC to transfer

Is this all right? If so, we should not be setting the `block_size` as a devicetree property. We should be using the `block_size` field from the `struct sdhc_data` provided for this request. Same thing for the other locations where we use a hardcoded `block-size`.

quytranpzz replied:
@danieldegrasse : just a little confirmation, those two have opposite meanings
- The `sector_count` is being used as the number of blocks for the DTC to transfer
- The `block_size `  is being used as the size of blocks for the DTC to transfer.

I see, I have removed the `block_size `setting as a devicetree property and used the `block_size` from struct `sdhc_data` instead. Thanks very much for letting me know.

danieldegrasse replied:
Sure, thanks for the clarification, sorry I missed that while looking at the DTC implementation

quytranpzz commented:
Rebases to update  HAL commit ID

danieldegrasse commented on drivers/sdhc/sdhc_renesas_ra.c:
```diff
@@ -0,0 +1,743 @@
+	r_sdhi_command_send_no_wait(p_ctrl, command, argument);
+
+	/* Wait for end of response, error or timeout */
+	return r_sdhi_wait_for_event(p_ctrl, SDHI_PRV_RESPONSE_BIT, timeout);
```
I'd still like to see this addressed- this implementation is polling on the status of the response bit: https://github.com/zephyrproject-rtos/hal_renesas/blob/31f1d1b6c78b6d1e2dbf7a63fdc71e7ba6a74740/drivers/ra/fsp/src/r_sdhi/r_sdhi.c#L2137

Does the hardware provide an interrupt that could be used here instead? If you'd prefer to keep this polling you may (it is your driver after all), but I'd prefer to use interrupts here if we can.

quytranpzz replied:
Our hardware doesn't provide a specific interrupt that could be used to check the response status bit. The response bit will combine all events for each command. May I keep the polling here?

danieldegrasse replied:
Of course- again, it is your driver, this was just a suggestion!

quytranpzz commented:
@danieldegrasse : Could you please revisit this PR? Thank you

pdgendt commented on drivers/sdhc/sdhc_renesas_ra.c:
```diff
@@ -0,0 +1,743 @@
+	return ret;
+}
+
+static const struct sdhc_driver_api sdhc_api = {.reset = sdhc_ra_reset,
+						.request = sdhc_ra_request,
+						.set_io = sdhc_ra_set_io,
+						.get_card_present = sdhc_ra_get_card_present,
+						.card_busy = sdhc_ra_card_busy,
+						.get_host_props = sdhc_ra_get_host_props};
```
Add the `DEVICE_API` wrapper and add a trailing comma, so it gets left aligned.

```suggestion
static DEVIC_API(sdhc, sdhc_api) = {
	.reset = sdhc_ra_reset,
	.request = sdhc_ra_request,
	.set_io = sdhc_ra_set_io,
	.get_card_present = sdhc_ra_get_card_present,
	.card_busy = sdhc_ra_card_busy,
	.get_host_props = sdhc_ra_get_host_props,
};
```

quytranpzz replied:
Updated, thanks

pdgendt commented on drivers/sdhc/sdhc_renesas_ra.c:
```diff
@@ -0,0 +1,743 @@
+	return ret;
+}
+
+static DEVICE_API(sdhc, sdhc_api) = {.reset = sdhc_ra_reset,
+				     .request = sdhc_ra_request,
+				     .set_io = sdhc_ra_set_io,
+				     .get_card_present = sdhc_ra_get_card_present,
+				     .card_busy = sdhc_ra_card_busy,
+				     .get_host_props = sdhc_ra_get_host_props};
```
Can you add the trailing comma so this gets nicely aligned with a single tab?

```suggestion
static DEVICE_API(sdhc, sdhc_api) = {
	.reset = sdhc_ra_reset,
	.request = sdhc_ra_request,
	.set_io = sdhc_ra_set_io,
	.get_card_present = sdhc_ra_get_card_present,
	.card_busy = sdhc_ra_card_busy,
	.get_host_props = sdhc_ra_get_host_props,
};
```

quytranpzz replied:
Sure,  I updated it, thank you

danieldegrasse reviewed (dismissed):
Thanks for all your work here!

thaoluonguw commented on boards/renesas/ek_ra8m1/ek_ra8m1.yaml:
```diff
@@ -10,3 +10,4 @@ toolchain:
 supported:
   - gpio
   - uart
+  - sdhc
```
@quytranpzz : Is it necessary to support as default?

thaoluonguw commented on drivers/sdhc/Kconfig.renesas_ra:
```diff
@@ -0,0 +1,13 @@
+	select USE_RA_FSP_DTC
+	select PINCTRL
+	help
+	  Enables Renesas Host controller driver
```
Should it be "Enable Renesas SD Host controller driver"?

quytranpzz replied:
Yes, I updated it.

thaoluonguw commented on drivers/sdhc/sdhc_renesas_ra.c:
```diff
@@ -0,0 +1,745 @@
+
+#include <zephyr/kernel.h>
+#include <zephyr/devicetree.h>
+#include <zephyr/drivers/pinctrl.h>
+#include <zephyr/drivers/gpio.h>
+#include <zephyr/irq.h>
+#include <soc.h>
+#include <zephyr/drivers/pinctrl.h>
```
Duplicated include <zephyr/drivers/pinctrl.h>

quytranpzz replied:
I removed it, thanks

duynguyenxa commented on boards/renesas/ek_ra8d1/ek_ra8d1-pinctrl.dtsi:
```diff
@@ -53,4 +53,23 @@
+			<RA_PSEL(RA_PSEL_SDHI, 4, 3)>,
+			<RA_PSEL(RA_PSEL_SDHI, 4, 4)>,
+			<RA_PSEL(RA_PSEL_SDHI, 4, 5)>,
+			<RA_PSEL(RA_PSEL_SDHI, 7, 0)>;
```
@quytranpzz This pin is conflicting with the spi0_default node

duynguyenxa commented on boards/renesas/ek_ra8d1/ek_ra8d1.dts:
```diff
@@ -160,3 +161,17 @@
 	pinctrl-names = "default";
 };
+
+&sdhc1 {
+	compatible = "renesas,ra-sdhc";
+	interrupt-names = "accs", "card", "dma-req";
+	interrupts = <60 12>, <61 12>, <62 12>;
+	pinctrl-0 = <&sdhc1_default>;
+	pinctrl-names = "default";
+	status = "okay";
+	sdmmc {
+		compatible = "zephyr,sdmmc-disk";
+		disk-name = "SD";
+		status = "okay";
+	};
+};
```
@thaoluonguw , @KhiemNguyenT , @soburi , the EK_RA8M1 and EK_RA8D1 doesn't support SD slot default on board but require external board, should we set its status as "okay" in board.dts?

quytranpzz replied:
I moved the sdhc support with the status "okay" into app overlay for EK_RA8M1 and EK_RA8D1 because it requires external hardware and avoids some pins conflicts as above, keeps only for MCK-RA8T1
@KhiemNguyenT @thaoluonguw @soburi @duynguyenxa: please take a look. Thank you

quytranpzz replied:
@danieldegrasse : sorry for inconvenience but could you please take another look? The driver implementation is unchanged, but the DTS support for the EK-RA8D1 and EK-RA8M1 boards has been updated as above discussed. Thank you

danieldegrasse replied:
So if I understand right, an external board is required for support here, which contains an SD card slot? A few questions:
- Is this external board standardized between multiple EVKs? That is, can it be reused on multiple Renesas EVKs?
- Is this board a product that can be purchased, or a custom rework to connect an SD card?

If the board is a product that can be purchased and is reusable across multiple boards, then I would suggest defining the board as a [shield](https://docs.zephyrproject.org/latest/hardware/porting/shields.html), so that it can be reused across multiple Renesas EVKs. You'll need to define a standard node name for the SDHC controller that exposes the pins the shield connects to- maybe something like `renesas_ext_sdhc`? See here for an example of this with NXP EVKs for an LCD interface: https://github.com/zephyrproject-rtos/zephyr/blob/2d60d248e846102867e4aca7a9e3d3e6ae591f06/boards/shields/lcd_par_s035/lcd_par_s035_8080.overlay#L33

quytranpzz replied:
@danieldegrasse : 
> Is this external board standardized between multiple EVKs? That is, can it be reused on multiple Renesas EVKs?

Yes, it's standardized and can be reused for multiple Renesas EVKs.
> Is this board a product that can be purchased, or a custom rework to connect an SD card?

This is a product by digilent and can be purchased ([pmod sd](https://digilent.com/reference/pmod/pmodsd/start?srsltid=AfmBOopc03O7OGG2h70JZGR_v8o20X7MKZBe7GgHKNkWXjVIIfs3lW7W))
So, I defined a board shield for Pmod SD and overlay the pin connection for EK-RA8M1 and EK-RA8D1 inside. 
Please take a look. Thank you very much.

quytranpzz committed:
tests: subsys: fs: Support tests/subsys/fs on RA8 boards

Add overlay property node and configs to support tests/subsys/fs/ext2,
tests/fs/fat_fs_api and tests/subsys/sd on RA8 boards

Signed-off-by: Quy Tran <quy.tran.pz@renesas.com>

quytranpzz committed:
boards: shields: Add support for pmod sd

Add support for Pmod SD and support overlay for EK-RA8M1, EK-RA8D1
to use with Pmod

Signed-off-by: Quy Tran <quy.tran.pz@renesas.com>

danieldegrasse commented on boards/shields/pmod_sd/boards/ek_ra8d1.overlay:
```diff
@@ -0,0 +1,29 @@
+	};
+};
+
+&pmod_sd_shield {
```
Can this be set at the board level, or does it need to be set within the shield overlay for this board?

quytranpzz replied:
I moved the pins into the pinctrl of the board.
For the RA8D1 shield overlay. I think it's better to set it here. It has pins conflicting with ethernet default onboard and the interrupt number was limited to enable it by default.

danieldegrasse commented on boards/shields/pmod_sd/boards/ek_ra8m1.overlay:
```diff
@@ -0,0 +1,23 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+&pinctrl {
```
Same question- can this pin control be defined at the board level?

quytranpzz replied:
I moved it into pin-ctrl at the board level.

danieldegrasse commented on boards/shields/pmod_sd/doc/index.rst:
```diff
@@ -0,0 +1,36 @@
+References
+**********
+
+- `Pmod SD product page`_
```
I think adding the PMOD pinout here might be useful as well:
![Screenshot 2024-12-19 at 12 08 46](https://github.com/user-attachments/assets/7dd88481-6ef4-4f8e-8ce0-962eb4fa5b2c)

quytranpzz replied:
Sure, I added PMOD pinout and futures for the document.

quytranpzz commented:
Rebases to solve conflicts
